### PR TITLE
No longer support checking Python 2

### PR DESCRIPTION
**Pull request by hauntsaninja**
Linking #12237

### CODE REVIEW

File: mypy/test/helpers.py [hunk 1]
```diff
@@ -287,6 +287,7 @@ def num_skipped_suffix_lines(a1: List[str], a2: List[str]) -> int:
 
 def testfile_pyversion(path: str) -> Tuple[int, int]:
     if path.endswith('python2.test'):
+        raise ValueError(path)
```
Comment (JelleZijlstra) [inline mypy/test/helpers.py:290]: Maybe just remove this case?

File: mypy/test/helpers.py [hunk 2]
```diff
@@ -296,6 +297,7 @@ def testfile_pyversion(path: str) -> Tuple[int, int]:
 
 def testcase_pyversion(path: str, testcase_name: str) -> Tuple[int, int]:
     if testcase_name.endswith('python2'):
+        raise ValueError(testcase_name)
```
Comment (JelleZijlstra) [inline mypy/test/helpers.py:300]: Same here

File: mypy/test/testdeps.py [hunk 3]
```diff
@@ -30,6 +30,7 @@ def run_case(self, testcase: DataDrivenTestCase) -> None:
         dump_all = '# __dump_all__' in src
         options = parse_options(src, testcase, incremental_step=1)
         if testcase.name.endswith('python2'):
+            raise ValueError(testcase.name)
```
Comment (JelleZijlstra) [inline mypy/test/testdeps.py:33]: And again

File: mypy/test/testparse.py [hunk 4]
```diff
@@ -32,6 +32,7 @@ def test_parser(testcase: DataDrivenTestCase) -> None:
     options = Options()
 
     if testcase.file.endswith('python2.test'):
+        raise ValueError(testcase.name)
```
Comment (JelleZijlstra) [inline mypy/test/testparse.py:35]: same

File: mypy/test/testpythoneval.py [hunk 5]
```diff
@@ -61,13 +57,7 @@ def test_python_evaluation(testcase: DataDrivenTestCase, cache_dir: str) -> None
     ]
     py2 = testcase.name.lower().endswith('python2')
```
Comment (JelleZijlstra) [inline mypy/test/testpythoneval.py:58]: again

**Review by hauntsaninja — COMMENTED** [review]
I left those in semi-intentionally, since it proves this removes all tests, before it removes the test scaffolding. But I can remove them

File: mypy/main.py [hunk 6]
```diff
@@ -951,6 +951,11 @@ def set_strict_flags() -> None:
     # The python_version is either the default, which can be overridden via a config file,
     # or stored in special_opts and is passed via the command line.
     options.python_version = special_opts.python_version or options.python_version
+    if options.python_version < (3,):
+        parser.error(
+            "Mypy no longer supports checking Python 2 code. "
+            "Consider pinning to mypy<=0.971 if you need to check Python 2 code."
```
Comment (sobolevn) [inline mypy/main.py:957]:
```suggestion
            "Consider pinning to mypy<0.980 if you need to check Python 2 code."
```

Comment (sobolevn) [reply mypy/main.py:957]: I think it is better to rely on a minor release here, not patch.

**Review by sobolevn — COMMENTED** [review]
There are still lots of python-2 specific logic pieces in `checker.py` and `checkmember.py` (probably somewhere else), but this is a great start!

File: test-data/unit/cmdline.test [hunk 7]
```diff
@@ -130,30 +130,6 @@ two/mod/__init__.py: note: See https://mypy.readthedocs.io/en/stable/running_myp
 two/mod/__init__.py: note: Common resolutions include: a) using `--exclude` to avoid checking one of them, b) adding `__init__.py` somewhere, c) using `--explicit-package-bases` or adjusting MYPYPATH
 == Return code: 2
 
-[case testFlagsFile]
-# cmd: mypy @flagsfile
```
Comment (JukkaL) [inline test-data/unit/cmdline.test:134]: There doesn't seem to be any other test case that uses `@flagsfile`, so this test case should not be deleted. We should just change it to use some other flags.

File: test-data/unit/cmdline.test [hunk 8]
```diff
@@ -130,30 +130,6 @@ two/mod/__init__.py: note: See https://mypy.readthedocs.io/en/stable/running_myp
...
--2
-main.py
-[file main.py]
-def f():
-    try:
-        1/0
-    except ZeroDivisionError, err:
-        print err
-
-[case testConfigFile]
```
Comment (JukkaL) [inline test-data/unit/cmdline.test:145]: Similar to above, this doesn't look Python 2 specific as such, but tests using `mypy.ini`. Please update the test case to use some other config file option.

File: test-data/unit/fine-grained-modules.test [hunk 9]
```diff
@@ -2158,23 +2158,6 @@ main.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mis
 main.py:2: error: Incompatible types in assignment (expression has type "int", variable has type "str")
 ==
 
-[case testMissingStubAdded2]
-# flags: --follow-imports=skip --py2
```
Comment (JukkaL) [inline test-data/unit/fine-grained-modules.test:2162]: This test case might still be relevant. I think that it's better to just drop the `--py2` flag but keep the test case.

**Comment by ilevkivskyi** [comment]
I think we should go ahead with this and fix `testFlagsFile` later separately. @hauntsaninja please don't forget to make a follow-up PR.

**Pull request merged**